src/spyglass/spikesorting/v1/recording.py

```diff
@@ -0,0 +1,730 @@
+                    "electrode_group_name"
+                ]  # reference for this electrode group
+                if (
+                    len(reference_electrode_group) == 1
+                ):  # unpack single reference
+                    reference_electrode_group = reference_electrode_group[0]
+                elif (int(sg_key["sort_reference_electrode_id"]) > 0) and (
+                    len(reference_electrode_group) != 1
+                ):
+                    raise Exception(
+                        f"Should have found exactly one electrode group for reference electrode,"
+                        f"but found {len(reference_electrode_group)}."
+                    )
+                if omit_ref_electrode_group and (
+                    str(e_group) == str(reference_electrode_group)
+                ):
+                    print(
+                        f"Omitting electrode group {e_group} from sort groups because contains reference."
+                    )
+                    continue
+                shank_elect = electrodes["electrode_id"][
+                    np.logical_and(
+                        electrodes["electrode_group_name"] == e_group,
+                        electrodes["probe_shank"] == shank,
+                    )
+                ]
+                if (
+                    omit_unitrode and len(shank_elect) == 1
+                ):  # omit unitrodes if indicated
+                    print(
+                        f"Omitting electrode group {e_group}, shank {shank} from sort groups because unitrode."
+                    )
+                    continue
+                cls.insert1(sg_key, skip_duplicates=True)
+                for elect in shank_elect:
+                    sge_key["electrode_id"] = elect
+                    cls.SortGroupElectrode.insert1(
+                        sge_key, skip_duplicates=True
+                    )
+                sort_group += 1
+
+
+@schema
+class SpikeSortingPreprocessingParameter(dj.Lookup):
+    definition = """
+    # Parameter for denoising (filtering and referencing/whitening) recording
+    # prior to spike sorting
+    preproc_param_name: varchar(200)
+    ---
+    preproc_param: blob
+    """
+
+    contents = [
+        [
+            "default",
+            {
+                "frequency_min": 300,  # high pass filter value
+                "frequency_max": 6000,  # low pass filter value
+                "margin_ms": 5,  # margin in ms on border to avoid border effect
+                "seed": 0,  # random seed for whitening
+            },
+        ]
+    ]
+
+    @classmethod
+    def insert_default(cls):
+        cls.insert1(cls.contents, skip_duplicates=True)
+
+
+@schema
+class SpikeSortingRecordingSelection(dj.Manual):
+    definition = """
+    # Raw voltage traces and parameter
+    -> Raw
+    -> SortGroup
+    -> IntervalList
+    -> SpikeSortingPreprocessingParameter
+    -> LabTeam
+    """
+
+
+@schema
+class SpikeSortingRecording(dj.Computed):
```

Comment:
I should probably wait for the tutorial, but what is the correct way to populate this table? Since `recording_id` is the primary key, but it isn't defined until inside of the make function.  This means that when I try to populate it with the `SpikeSortingRecordingSelection` key I get this error:

> DataJointError: A table must have dependencies from its primary key for auto-populate to work

Reply:
My thought was that this table would be populated by calling the `make` method (i.e. `SpikeSortingRecording.populate()` which calls `make`) like any other `dj.Computed` table. Did that not work?

Reply:
Not when I try it.  I'm pretty sure that for that to work `SpikeSortingRecordingSelection` needs to be a primary key (which I think is the normal standard). Other wise when you call populate it has nothing to start from.  For example, after populating `SpikeSortingRecordingSelection`, when I run this

`SpikeSortingRecording().populate()`

I get this error

> DataJointError: A table must have dependencies from its primary key for auto-populate to work

Reply:
Ah OK, my mistake. Will fix this. Maybe I have to introduce `recording_id` at the selection table.

Reply:
Introduced new `insert_selection` method to generate `recording_id`. I added docstring to the table definition so that users would know to use it. The motivation for doing this is to avoid the problem of having a very large composite primary key downstream. Let me know if this solution seems satisfactory to you.